state.go

```diff
@@ -584,19 +585,75 @@ func NewUpdateErrorState(err menderError, update UpdateResponse) State {
 }
 
 func (ue *UpdateErrorState) Handle(ctx *StateContext, c Controller) (State, bool) {
-	// TODO error handling, repeat status update for as long as needed
-	c.ReportUpdateStatus(ue.update, statusFailure)
-
-	// TODO fetch logs from logger
-	c.UploadLog(ue.update, []LogEntry{
-		LogEntry{
-			Timestamp: time.Now().Format(time.RFC3339),
-			Level:     "error",
-			Message:   "update failed",
-		},
-	})
+	return NewUpdateStatusReportState(ue.update, statusFailure), false
+}
+
+// Wrapper for mandatory update state reporting, the state handler will attempt
+// to report state for a number of times. In case of recurring failure, the
+// update is deemed as failed.
+type UpdateStatusReportState struct {
+	CancellableState
+	update UpdateResponse
+	status string
+}
+
+func NewUpdateStatusReportState(update UpdateResponse, status string) State {
+	return &UpdateStatusReportState{
+		CancellableState: NewCancellableState(BaseState{
+			id: MenderStateUpdateStatusReport,
+		}),
+		update: update,
+		status: status,
+	}
+}
+
+func (usr *UpdateStatusReportState) Handle(ctx *StateContext, c Controller) (State, bool) {
+	if err := StoreStateData(ctx.store, StateData{
+		Id:           usr.Id(),
+		UpdateInfo:   usr.update,
+		UpdateStatus: usr.status,
+	}); err != nil {
+		log.Errorf("failed to store state data in update status report state: %v",
+			err)
+		// TODO: what now?
```

Comment:
Makes sense. Since we don't support rollback at the moment, I'll update the comment to reflect that.